parsers/US_PJM.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+
+"""Parser for the PJM area of the United States."""
+
+import arrow
+from bs4 import BeautifulSoup
+import demjson
+import re
+import requests
+
+
+url = 'http://www.pjm.com/markets-and-operations.aspx'
+
+mapping = {
+            'Coal': 'coal',
+            'Gas': 'gas',
+            'Hydro': 'hydro',
+            'Multiple Fuels': 'unknown',
+            'Nuclear': 'nuclear',
+            'Oil': 'oil',
+            'Other': 'unknown',
+            'Other Renewables': 'unknown',
+            'Solar': 'solar',
+            'Wind': 'wind'
+            }
+
+
+def extract_data(session = None):
+    """
+    Makes a request to the PJM data url.
+    Finds timestamp of current data and converts into a useful form.
+    Finds generation data inside script tag.
+    Returns a tuple of generation data and datetime.
+    """
+
+    s = session or requests.Session()
+    req = requests.get(url)
+    soup = BeautifulSoup(req.content, 'html.parser')
+
+    try:
+        time_div = soup.find("div", id="asOfDate").text
+    except AttributeError:
+        raise LookupError('No data is available for US-PJM.')
+
+    time_pattern = re.compile(r"""(\d{1,2}     #Hour can be 1/2 digits.
```

Comment:
why use a dateformat and parse using `datetime` or `arrow`?

Reply:
The time format has some weirdness in it (a.m./p.m.) which makes it harder to parse. I think string operations would be needed to get it into a workable form and the regex is pretty clear on what it's searching for. However there a few problems with the datetime already so I'll see if it can be simplified.

Reply:
@systemcatch have you been able to find a "standalone" page, url or csv for the real-time Generation Fuel Mix, that is not embedded within the "Markets & Operations" page?

I searched for hours, even creating an account for their "PJM Tools" https://esuite.pjm.com/mui/index.htm system data accessibility applications, yet couldn't come up with anything there either. :(

So I wil be fascinated to see how you guys put this parser together! :)

FYI - for NYISO, timestamps with 00:00 --> 00:59:59 represents from 12:00am --> 12:59:59am

Reply:
noted @systemcatch